packages/panzer/mini-em/src/closures/MiniEM_RandomScalarForcing_impl.hpp

```diff
@@ -0,0 +1,83 @@
+#ifndef MINIEM_RANDOM_SCALAR_FORCING_IMPL_HPP
+#define MINIEM_RANDOM_SCALAR_FORCING_IMPL_HPP
+
+#include <cmath>
+
+#include "Panzer_BasisIRLayout.hpp"
+#include "Panzer_Workset.hpp"
+#include "Panzer_Workset_Utilities.hpp"
+#include "Panzer_GatherBasisCoordinates.hpp"
+
+#include "Panzer_Traits.hpp"
+#include "Kokkos_Random.hpp"
+#include "MiniEM_Sacado_Kokkos_Random.hpp"
+#include "Kokkos_ArithTraits.hpp"
+
+namespace mini_em {
+
+//**********************************************************************
+template <typename EvalT,typename Traits>
+RandomScalarForcing<EvalT,Traits>::RandomScalarForcing(const std::string & name,
+                                                       const panzer::IntegrationRule & ir,
+                                                       const panzer::FieldLayoutLibrary & fl,
+                                                       const unsigned int & seed,
+                                                       const double & rangeMin,
+                                                       const double & rangeMax,
+                                                       const std::string& basisName)
+{
+  using Teuchos::RCP;
+
+  Teuchos::RCP<PHX::DataLayout> data_layout = ir.dl_scalar;
+  ir_degree = ir.cubature_degree;
+  ir_dim = ir.spatial_dimension;
+
+  current = PHX::MDField<ScalarT,Cell,Point>(name, data_layout);
+  this->addEvaluatedField(current);
+
+  Teuchos::RCP<const panzer::PureBasis> basis = fl.lookupBasis(basisName);
+  const std::string coordName = panzer::GatherBasisCoordinates<EvalT,Traits>::fieldName(basis->name());
+  coords = PHX::MDField<const ScalarT,Cell,Point,Dim>(coordName, basis->coordinates);
+  this->addDependentField(coords);
+
+  std::string n = "Random Scalar Forcing";
+  this->setName(n);
+
+  rangeMin_ = rangeMin;
+  rangeMax_ = rangeMax-rangeMin;
+  seed_ = seed;
+}
+
+//**********************************************************************
+template <typename EvalT,typename Traits>
+void RandomScalarForcing<EvalT,Traits>::evaluateFields(typename Traits::EvalData workset)
+{
+  using panzer::index_t;
+
+  // double time = workset.time;
+  auto current_h = Kokkos::create_mirror_view(current.get_static_view());
+
+  if (ir_dim == 3) {
+    for (index_t cell = 0; cell < workset.num_cells; ++cell) {
+      for (int point = 0; point < current_h.extent_int(1); ++point) {
```

Comment:
These functions are not on device. Should they be moved there for efficiency?

Reply:
And yes!